### PR TITLE
#167165017 Creates Signup/Login Page

### PR DESCRIPTION
**Pull request by OdunayoOkebunmi**
### What does this PR do?
Creates the signup and login page
Add a signup route
Add a login route
Writes tests for newly added  component

#### Description of Task to be completed?
 Users should be able to Signup and or Register using Sign up forms on the signup route.
 Users should be able to login using signin forms on the signup route.

#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin ft-user-signup-167165017

# install dependencies
$ npm install

# unit and integration tests are available by running this script
$ npm test

# start the development server
$ npm run dev
```

After doing this, you should see a message appear on the console, with the app running on Port 8080, you can view the app on `localhost:8080`. You can view the app on the above stated route

#### Any background context you want to provide?
N/a

#### What are the relevant pivotal tracker stories?
[#167165017](https://www.pivotaltracker.com/story/show/167165017)

#### Screenshots (if appropriate)
Screenshot 2019-08-29 at 22 50 27
Screenshot 2019-08-29 at 22 51 38

### CODE REVIEW

File: src/components/common/Form/Form.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable react/prop-types */
```
Comment (mazma1) [inline src/components/common/Form/Form.js]: Why do we have this `disable` statement?

File: src/components/common/Form/Form.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const FormContainer = ({ onSubmit, children }) => (
```
Comment (mazma1) [inline src/components/common/Form/Form.js]: On your general naming conventions:

1. Should folders be in pascal case?
2. Should components exporting functions or classes have PascalCase filenames in both cases?
3. Should your component names match their corresponding filenames?

Find out the correct conventions and make corrections where necessary

File: src/components/common/isEmpty.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+const isEmpty = value => (
+  value === undefined
+    || value === null
+    || (typeof value === 'object' && Object.keys(value).length === 0)
+    || (typeof value === 'string' && value.trim().length === 0)
+);
+
+export default isEmpty;
```
Comment (mazma1) [inline src/components/common/isEmpty.js]: Is this a component? Why is it in the components folder?

File: src/components/views/auth/SignUp/SignUp.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
+import React, { useState, Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import Toastr from 'toastr';
+import { registerAccount } from '../../../../redux/actions/authActions';
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: Can we get rid of all the relative import hell `../../../../../../../../`?

Webpack's `alias` similar to what was described [here](https://medium.com/@kkomaz/another-solution-to-absolute-path-setup-in-create-react-app-cccb38d8eea8) should help

File: src/components/views/auth/SignUp/SignUp.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+import InputField from '../../../common/Form/InputField';
+import LeftDiv from '../commons/leftDiv';
+import SocialButtons from '../commons/buttons';
+import { signupText } from '../commons/leftDivText';
+import '../commons/auth.scss';
+
+const SignUp = (props) => {
+  // const { errors } = props;
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: Delete any code comment that is not needed anymore

File: src/components/views/auth/SignUp/SignUp.js [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  const [firstName, updateFirstName] = useState('');
+  const [lastName, updateLastName] = useState('');
+  const [email, updateEmail] = useState('');
+  const [password, updatePassword] = useState('');
+  const [confirmPassword, updateConfirmPassword] = useState('');
+
+  const registerUser = async (e) => {
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: `e`? Descriptive variable names please

File: src/components/views/auth/commons/buttons.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+      </button>
+    </div>
+    <div className="input-field">
+      <button className="btn--block btn-facebook" type="submit" value="Sign in">
+        <span className="social-icon-left"><FontAwesomeIcon icon={faFacebook} /></span>
+        Continue with Facebook <span className="social-icon-right"><FontAwesomeIcon icon={faArrowRight} /></span>
+      </button>
+    </div>
+  </div>
+
```
Comment (mazma1) [inline src/components/views/auth/commons/buttons.js]: This is not a good abstraction. `<Button />` currently has a repetition - `12 - 18` and `19 -24`. Also, it is not reusable as it is strictly hardcoded for social logins.

Can we have more flexible `<Button />` component?

File: src/components/views/auth/SignUp/SignUp.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    if (password !== confirmPassword) {
+      Toastr.error('Passwords do not match');
+      return;
+    }
+    await props.registerAccount(userData, props.history);
+  };
+  return (
+    <Fragment>
+      <div className="container-body">
+        <LeftDiv signinText={signupText} />
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: Why is `signinText` = `signupText`?

File: src/components/views/auth/commons/leftDiv.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable react/prop-types */
+import React from 'react';
+
+const LeftDiv = ({ signinText }) => (
```
Comment (mazma1) [inline src/components/views/auth/commons/leftDiv.js]: Is `<leftDiv />` used by only the signin component?

File: src/components/views/auth/signIn/Signin.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+  const userData = {
+    email,
+    password,
+  };
+
+
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+    onSubmit(userData, props.history);
```
Comment (mazma1) [inline src/components/views/auth/signIn/Signin.js]: Can this function be better named so it easy to tell what it does? `onSubmit` is a generic event listener

File: src/config/API.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+const API_URL = process.env.NODE_ENV === 'development' ? 'http://localhost:3000' : 'mazus-ah-staging.herokuapp.com';
+
```
Comment (mazma1) [inline src/config/API.js]: Can the app URLs be moved to env variables so we don't have to edit this file if it has to change?

File: src/redux/actions/authActions.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    loading: false,
+    error,
+  },
+});
+
+
+export const signIn = (data, history) => async (dispatch) => {
+  dispatch(authLoading());
+  try {
+    const logInUser = await axios.post(`${AUTH_API}/signin`, data);
```
Comment (mazma1) [inline src/redux/actions/authActions.js]: Can we have an API service that abstracts `axios.post()`, `axios.get()`, etc?

File: src/components/common/setAuthToken.js [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+
+const setAuthToken = (token) => {
+  if (token) {
+    axios.defaults.headers.common.Authorization = token;
+  } else {
+    delete axios.defaults.headers.common.Authorization;
+  }
+};
+
+export default setAuthToken;
```
Comment (mazma1) [inline src/components/common/setAuthToken.js]: I mention creating an API service somewhere else. I think adding token to the header can move in there as well. You can look up `axios.interceptors`

File: src/components/views/auth/SignUp/SignUp.js [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+};
+SignUp.defaultProps = {
+  errors: {},
+  auth: {},
+};
+const mapStateToProps = state => ({
+  auth: state.auth,
+  errors: state.errors,
+  loading: state.loading,
+});
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: From your state, shouldn't you have:
```
errors: state.auth.error,
loading: state.auth.loading,
```
Screenshot 2019-08-30 at 12 50 06 AM

File: src/components/views/auth/SignUp/SignUp.js [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+                value={confirmPassword}
+                change={e => updateConfirmPassword(e.target.value)}
+                name="confirmPassword"
+                placeholder="Confirm Password"
+                className="form-input"
+                required
+              />
+            </div>
+            <div className="input-field">
+              <button className="btn--block bttn-primary btn-submit" type="submit" value="Sign up"> Create your free account<i className="arrow-forward material-icons">arrow_forward</i></button>
```
Comment (mazma1) [inline src/components/views/auth/SignUp/SignUp.js]: I don't know if sign up currently works but I got this error when I tried:

Screenshot 2019-08-30 at 12 49 25 AM

Also, consider disabling the button when a signup request is still loading.

File: src/components/views/auth/SignUp/SignUp.js (hunk 14 above)
Comment (Tunji17) [reply src/components/views/auth/SignUp/SignUp.js]: In the initial state object, we named it errors and loading, that's why we have state.errors and not state.auth.errors

File: src/components/common/auth/leftDiv.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const LeftDiv = ({ Text }) => (
```
Comment (mazma1) [inline src/components/common/auth/leftDiv.js]: Why does `Text` begin with a capital letter?

File: src/components/views/auth/signIn/SignIn.spec.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
+/* eslint-disable no-unused-vars */
+import React from 'react';
+import { shallow, mount } from 'enzyme';
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import { BrowserRouter as Router } from 'react-router-dom';
+import Signin, { SigninComponent } from './Signin';
+import store from '../../../../redux/store/index';
```
Comment (mazma1) [inline src/components/views/auth/signIn/SignIn.spec.js]: Thought I pointed out this relative import thing?

File: src/components/views/auth/signIn/Signin.js [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+
+const Signin = (props) => {
+  const [values, setValues] = useState({
+    email: '',
+    password: '',
+  });
+  const { loading, history, onSubmit } = props;
+
+  const onChange = (e) => {
+    e.persist();
```
Comment (mazma1) [inline src/components/views/auth/signIn/Signin.js]: Can all occurrences of this `e` variable be properly named?

File: src/components/common/auth/leftDiv.js (hunk 16 above)
Comment (Tunji17) [reply src/components/common/auth/leftDiv.js]: okay I would correct it

File: src/app.js [hunk 19]
```diff
@@ -3,22 +3,21 @@ import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
 import { Provider } from 'react-redux';
 import store from './redux/store';
 import './app.scss';
-import Welcome from './components/views/Welcome/Welcome';
-import NotFound from './components/views/NotFound/NotFound';
-import Header from './components/common/header';
+import Welcome from './components/views/welcome/Welcome';
+import SignIn from './components/views/auth/signIn/Signin';
+import SignUp from './components/views/auth/signUp/SignUp';
+import NotFound from './components/views/notFound/NotFound';
```
Comment (mazma1) [inline src/app.js:9]: Change these to absolute paths, whether it is deeply nested or not

File: src/components/common/auth/leftDiv.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const LeftDiv = ({ text }) => (
+  <div className="left-block">
+    <div className="img-field">
+      <img src="https://res.cloudinary.com/mazus/image/upload/c_scale,w_313/v1567007966/athors_haven_fkkrsv.png" alt="Author's Haven Logo" />
```
Comment (mazma1) [inline src/components/common/auth/leftDiv.js:7]: The link to the logo should be moved to `.env`, it's repeated a lot

File: src/components/views/auth/signUp/SignUp.js [hunk 21]
```diff
@@ -18,6 +20,16 @@ const SignUp = (props) => {
     password: '',
     confirmPassword: '',
   });
+  useEffect(() => {
+    const { location, socialSignOn, history } = props;
+    const tokenString = location && location.search;
+    if (tokenString) {
+      const token = getToken(tokenString);
+      socialSignOn(token);
+      history.push('/');
```
Comment (mazma1) [inline src/components/views/auth/signUp/SignUp.js:28]: What happens if `socialSignOn(token)` fails, will the user still be redirected to `/`?

File: src/components/views/auth/signUp/SignUp.js [hunk 22]
```diff
@@ -18,6 +20,16 @@ const SignUp = (props) => {
     password: '',
     confirmPassword: '',
   });
+  useEffect(() => {
```
Comment (mazma1) [inline src/components/views/auth/signUp/SignUp.js:22]: What does this hook do?

File: src/redux/reducers/__tests__/socialReducer.spec.js [hunk 23]
```diff
@@ -0,0 +1,28 @@
+/* eslint-disable import/no-duplicates */
+import { getCurrentUser, getError } from '@Redux/actions/socialActions'; 
+import { initialState } from '../socialReducer';
+import socialReducer from '../socialReducer';
```
Comment (mazma1) [inline src/redux/reducers/__tests__/socialReducer.spec.js:4]: Can we be consistent with import statements absolute paths?

File: src/redux/actions/socialActions.js [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+
+export const getError = error => ({
+  type: GET_ERRORS,
+  payload: error,
+});
+export const socialAuthPath = token => async (dispatch) => {
+  try {
+    localStorage.setItem('token', token);
+    const decoded = jwtDecode(token);
+    dispatch(getCurrentUser(decoded));
```
Comment (mazma1) [inline src/redux/actions/socialActions.js]: `getCurrentUser` sounds like you're making another request to the database..

File: src/redux/actions/socialActions.js [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+export const getToken = (tokenString) => {
+  const startIndex = tokenString.indexOf('=') + 1;
+  const token = tokenString.slice(startIndex);
+  return token;
+};
+
+export const socialLink = (type) => {
+  window.location.href = `${process.env.API_BASE_URL}/auth/${type}`;
+};
```
Comment (mazma1) [inline src/redux/actions/socialActions.js:31]: Your variable names seem a bit off, I'll just make few suggestions that show intent:
- GET_ERRORS => SOCIAL_LOGIN_FAILURE
- AUTH_SOCIAL => SOCIAL_LOGIN_SUCCESS
- getCurrentUser => setCurrentError
- socialAuthPath => authorizeSocialUser
- getError => (find something for this guy)

Makes sense? You can even come up with better names that clearly tells what each thing does

File: src/redux/actions/socialAction.spec.js [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+    moxios.install();
+  });
+
+  afterEach(() => {
+    moxios.uninstall();
+  });
+
+  it('creates SET_CURRENT_USER when social login is successful', () => {
+    const { socialAuthPath } = actions;
+    expect(typeof socialAuthPath(authResponse)(dispatch)).toEqual('object');
```
Comment (mazma1) [inline src/redux/actions/socialAction.spec.js]: After checking that it is an object, what next?

File: src/redux/actions/socialAction.spec.js [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+  });
+
+  it('should return a decode object', () => {
```
Comment (mazma1) [inline src/redux/actions/socialAction.spec.js]: it should return a decode object when what happens?

File: src/redux/actions/socialAction.spec.js [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+    const decoded = { id: 1, email: 'example@email.com' };
+    const action = actions.getCurrentUser(decoded);
+    expect(action).toEqual({ type: 'AUTH_SOCIAL', payload: decoded });
+  });
+  it('should return an error', () => {
```
Comment (mazma1) [inline src/redux/actions/socialAction.spec.js]: it should return an error when what happens?

File: src/redux/actions/socialAction.spec.js [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+const dispatch = jest.fn(() => ({
+  type: AUTH_SOCIAL,
+  payload: {
+    user: authResponse,
+  },
+}));
+
+describe('Action tests', () => {
```
Comment (mazma1) [inline src/redux/actions/socialAction.spec.js]: "Action tests", are you testing the whole actions?

File: src/components/views/auth/signUp/SignUp.spec.js [hunk 30]
```diff
@@ -85,4 +88,20 @@ describe('SignUp Component', () => {
...
+      search: 'localhost:8080/signup?token=abc',
+    };
+    mount(
+      <Provider store={reduxStore}>
+        <Router>
+          <SignUp location={location} history={history} />
+        </Router>
+      </Provider>,
+    );
+    expect(history.push).toHaveBeenCalled();
```
Comment (mazma1) [inline src/components/views/auth/signUp/SignUp.spec.js]: Why are you not checking how many times `history.push` is called?

File: src/components/views/auth/signUp/SignUp.spec.js [hunk 31]
```diff
@@ -85,4 +88,20 @@ describe('SignUp Component', () => {
     form.simulate('submit');
     expect(registerUser).toHaveBeenCalledTimes(0);
   });
+  it('should successful redirect user to dashboard on successful mount', () => {
```
Comment (mazma1) [inline src/components/views/auth/signUp/SignUp.spec.js]: What should successful redirect user to dashboard on successful mount?

File: src/components/views/auth/signUp/SignUp.js [hunk 32]
```diff
@@ -128,21 +140,25 @@ SignUp.propTypes = {
...
 
 const mapStateToProps = state => ({
   auth: state.auth,
   errors: state.auth.errors,
   loading: state.auth.loading,
+  social: state.social,
 });
 const mapDispatchToProps = dispatch => ({
   onSubmit: (newUser, history) => dispatch(registerAccount(newUser, history)),
+  socialSignOn: token => dispatch(socialAuthPath(token)),
```
Comment (mazma1) [inline src/components/views/auth/signUp/SignUp.js]: This is the problem with variable names that are not descriptive(`socialAuthPath`), I couldn't tell what the dispatched action does just by the name

**Pull request closed without merging**